src/client/Microsoft.Identity.Client/ClientApplicationBase.cs

```diff
@@ -174,18 +174,49 @@ private async Task<IEnumerable<IAccount>> GetAccountsInternalAsync(ApiIds apiId,
                             StringComparison.OrdinalIgnoreCase));
                 }
 
-                foreach (IAccount account in brokerAccounts)
+                brokerAccounts = await FilterBrokerAccountsByEnvAsync(brokerAccounts).ConfigureAwait(false);
+                return brokerAccounts;
+            }
+
+            return Enumerable.Empty<IAccount>();
+        }
+
+        private async Task<IEnumerable<IAccount>> FilterBrokerAccountsByEnvAsync(IEnumerable<IAccount> brokerAccounts)
+        {
+            ServiceBundle.DefaultLogger.Verbose($"Filtering broker accounts by env. Before fitering: " + brokerAccounts.Count());
+
+            ISet<string> allEnvs = new HashSet<string>(
+                brokerAccounts.Select(aci => aci.Environment),
+                StringComparer.OrdinalIgnoreCase);
+
+            var instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                Authority,
+                allEnvs,
+                CreateRequestContext(Guid.NewGuid())).ConfigureAwait(false);
+
```

Comment:
nit: is there a reason to create a new guid here and not use the RC from the service bundle?